Add commands `/ back` `/exit` for `/quest `

NikitaZavorin opened the pull request.

NikitaZavorin committed:
new branch for test back and exit buttons

NikitaZavorin committed:
new branch for test back and exit buttons

NikitaZavorin committed:
adding new commands `/exit` and `/back`

NikitaZavorin committed:
adding new commands `/exit` and `/back`

NikitaZavorin committed:
remove unused keyboard imported from aiogram.utils

NikitaZavorin committed:
remove unused keyboard imported from aiogram.utils

NikitaZavorin committed:
checking `/start` command in user_input for quitting state

NikitaZavorin committed:
checking `/start` command in user_input for quitting state

NikitaZavorin committed:
fixing expression

NikitaZavorin committed:
fixing expression

NikitaZavorin committed:
refactor department -> user_input

NikitaZavorin committed:
refactor department -> user_input

wak3upalex commented on bot/buttons.py:
```diff
@@ -92,6 +98,11 @@ async def set_main_menu(bot:Bot):
         BotCommand(command='/quest',description= 'Пройди квест для определения своего отдела'),
+
+        #удалить при надобности
+        BotCommand(command='/back', description='Возвращает в начало квеста'),
+        BotCommand(command='/exit', description='Выходит из состояния квеста '),
```
@NikitaZavorin Is it possible to show these commands only when user is in question section?

NikitaZavorin replied:
I'm not sure, but I can try. Or you can add some transparent hints for the user. I have attached an example

![image](https://github.com/user-attachments/assets/b8b4e766-4a73-443f-919e-a2ca91c58b7c)

wak3upalex replied:
This was fixed in [this resolved conversation](https://github.com/wak3upalex/eestecbot/pull/35#discussion_r1780105416)

wak3upalex committed:
Hide reply buttons after `/start` command

wak3upalex commented on bot/inLineButtons_quest.py:
```diff
@@ -138,9 +138,43 @@ async def handle_answer(message: types.Message, state: FSMContext, next_state: S
+
+    if user_input == "/start":
+        print(f"Пользователь вышел из теста {message.from_user.id}: {user_input}.")
+        await message.answer(
+            text='Вы вышли из прохождения теста командой старт',
+            reply_markup=ReplyKeyboardRemove(),
+        )
```
@NikitaZavorin I added parameter to hide buttons for answer after `\start` command during `quest`, because there was a problem like this
<img width="496" alt="image" src="https://github.com/user-attachments/assets/cea7041e-ec02-4116-8d48-3c409cde9a23">

wak3upalex reviewed (commented):
Please check my question in comment on lines 103 and 104

NikitaZavorin committed:
Now, only when the user is in the /quest` command, he sees only the `/back` and `/exit` commands in the prompts

NikitaZavorin committed:
Merge remote-tracking branch 'origin/quest-back-exit-commands' into quest-back-exit-commands

# Conflicts:
#	bot/inLineButtons_quest.py

NikitaZavorin committed:
Now, only when the user is in the /quest` command, he sees only the `/back` and `/exit` commands in the prompts

NikitaZavorin committed:
Now, only when the user is in the /quest` command, he sees only the `/back` and `/exit` commands in the prompts

wak3upalex committed:
Revert "Now, only when the user is in the /quest` command, he sees only the `/back` and `/exit` commands in the prompts"

This reverts commit a5dbed97a09ea73b85ceb79933de43a6582b9fe5.

wak3upalex committed:
Revert "refactor department -> user_input"

This reverts commit bfdea5c046572b36077a8f505ebbd7fc03f4ed4e.

wak3upalex committed:
Revert "fixing expression"

This reverts commit d795eae687d5c3f9026210bd43917dcd270d2d72.

wak3upalex committed:
Revert "checking `/start` command in user_input for quitting state"

This reverts commit 625df6f015019473f5339921e1f5763cead0d0b8.

wak3upalex committed:
Revert "remove unused keyboard imported from aiogram.utils"

This reverts commit faed96e0b052119373ca3a0f48a95fb240d2d007.

wak3upalex committed:
Revert "adding new commands `/exit` and `/back`"

This reverts commit 105d6cd636bfe9d73de7d2c4d27817ac044cd4c2.

NikitaZavorin committed:
Merge remote-tracking branch 'origin/quest-back-exit-commands' into quest-back-exit-commands

# Conflicts:
#	bot/buttons.py

NikitaZavorin committed:
Now, only when the user is in the `/quest` command, he sees only the `/back` and `/exit` commands in the prompts

NikitaZavorin committed:
refactor department -> user_input

NikitaZavorin committed:
fix `/exit` command and i added that if you write the `/start` command during the passage of `/quest`, then you will be returned

wak3upalex commented on bot/inLineButtons_quest.py:
```diff
@@ -116,8 +116,31 @@ class TestStates(StatesGroup):
+dp.startup.register(set_main_menu)
+
+
+async def quest_menu(bot: Bot):
+    main_menu_commands = [
+        BotCommand(command='/back', description='Возвращает в начало квеста'),
+        BotCommand(command='/exit', description='Выходит из состояния квеста')
+    ]
+    await bot.set_my_commands(main_menu_commands)
```
Is it being used? I thought we decided not to use "special" menu for `/quest` as it is impossible to refresh menu

NikitaZavorin replied:
I was able to implement two menus in our code

wak3upalex replied:
Okay. I tested it. On desktop version for MacOS. It doesn't change menu **immediately**, but after quitting chat and entering again it works
<img width="620" alt="image" src="https://github.com/user-attachments/assets/d5cb7eaf-0075-4840-af3e-e8b16e98722f">
<img width="620" alt="image" src="https://github.com/user-attachments/assets/74086184-a6f0-4549-b8d7-099a99cfe540">

wak3upalex replied:
> I was able to implement two menus in our code

Yeah. I see it now. It works not super correctly. Buuuuuut I think it is because of the Telegram working process

wak3upalex commented on bot/inLineButtons_quest.py:
```diff
@@ -138,9 +161,52 @@ async def handle_answer(message: types.Message, state: FSMContext, next_state: S
+        print(f"Пользователь вышел из теста {message.from_user.id}: {user_input}.")
+        await set_main_menu(bot)
+        await state.clear()
+        await message.answer("Вы вышли из прохождения теста командой старт\nОцените продуманность кода по 10 бальной шкале\nНапишите цифру от 1 до 10.")
```
I don't see where user's answer is being saved? And I tested this and message with rate was not handled
```console
2024-09-29 21:17:13,873 : INFO : aiogram.event : Update id=427536492 is not handled. Duration 1 ms by bot id=7819575371
```

NikitaZavorin replied:
I have everything displayed, I will attach screenshots
<img width="1027" alt="image" src="https://github.com/user-attachments/assets/4ae3c816-96e8-4c9b-9b59-d81ca874860d">
<img width="574" alt="image" src="https://github.com/user-attachments/assets/644b239d-c46e-4e20-9704-d402f65fa94d">

wak3upalex replied:
Yeah. Here you are right, but if you leave a number as a rate for our code
Here you can see (especially in time of message being sent and its log time) After rate – bot doesn't answer and just being quiet
<img width="1439" alt="image" src="https://github.com/user-attachments/assets/9e0f8eea-e7c6-4862-88b7-f011cf4ac88e">

NikitaZavorin replied:
I understand you, I will add processing of user ratings in the next update

wak3upalex replied:
Currently I will comment this feature for later work. But this can be a good concept for future feedback collection! Thanks!

wak3upalex committed:
Hide `ReplyKeyboard` after `/start` command

wak3upalex reviewed (commented):
@NikitaZavorin Please respond to my questions ASAP

wak3upalex committed:
Changed message for `/quest`

Now additional commands for `/quest`: `/back`,`/exit` are mentioned and describe in the first message for the user for better user experience

wak3upalex committed:
Hide reply after `/start`

Hiding reply about rating `/start` handler in quest. Possibly will be implemented ihn future and currently works incorrectly, and it is better to hide it to avoid user's confusions

wak3upalex reviewed (approved):
Good

wak3upalex commented on bot/inLineButtons_quest.py:
```diff
@@ -167,7 +167,10 @@ async def handle_answer(message: types.Message, state: FSMContext, next_state: S
-        await message.answer("Для того чтобы заново начать квест напишите /quest.")
+        await message.answer(
+            text="Для того чтобы заново начать квест напишите /quest.",
+            reply_markup=ReplyKeyboardRemove()
```
Hiding buttons for answer

wak3upalex reviewed (approved):
Finally. I will squah commits into more compact one commit
